lib/posix/pthread.c

```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
  *
  * See IEEE 1003.1
  */
-int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
-		   void *(*threadroutine)(void *), void *arg)
+int pthread_create(pthread_t *th, const pthread_attr_t *_attr, void *(*threadroutine)(void *),
+		   void *arg)
 {
-	int rv;
-	int32_t prio;
+	int err;
 	k_spinlock_key_t key;
-	uint32_t pthread_num;
-	k_spinlock_key_t cancel_key;
-	pthread_condattr_t cond_attr;
-	struct posix_thread *thread;
-	const struct pthread_attr *attr = (const struct pthread_attr *)_attr;
+	struct posix_thread *t;
+	pthread_barrier_t barrier;
+	struct pthread_attr attr;
 
-	/*
-	 * FIXME: Pthread attribute must be non-null and it provides stack
-	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
-	 * attrib as NULL but zephyr needs it initialized with valid stack.
-	 */
-	if ((attr == NULL) || (attr->initialized == 0U)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (!pthread_attr_is_valid((const struct pthread_attr *)_attr)) {
 		return EINVAL;
 	}
 
-	key = k_spin_lock(&pthread_pool_lock);
-	for (pthread_num = 0;
-	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
-		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
-			break;
+	/* copy the attribute in to the */
+	(void)memcpy(&attr, _attr, sizeof(attr));
+
+	key = k_spin_lock(&q_lock);
+	/* recycle old nodes if they are available */
+	do {
```

Comment:
What is this outer do loop for? Just to declare safe_t? Is this a common pattern?

Reply:
Just an extra scope, because we don't need `safe_t` after this point.

Reply:
Sure, welcome to C, I guess. Could just use braces instead of a do though; I was a bit confused by the 'continue' inside the inner loop and wondering if you intended that to go back to the top of the 'do {' loop.